File: websocket-platform/framework/websocket/message.go
package websocket

import (
	"encoding/json"
	"time"
)

// MessageType 消息类型
type MessageType string

const (
	MessageTypeText         MessageType = "text"         // 文本消息
	MessageTypeFile         MessageType = "file"         // 文件消息
	MessageTypeToolCall     MessageType = "tool_call"    // 工具调用
	MessageTypeToolResult   MessageType = "tool_result"  // 工具结果
	MessageTypeHeartbeat    MessageType = "heartbeat"    // 心跳消息
	MessageTypeStatus       MessageType = "status"       // 状态消息
	MessageTypeError        MessageType = "error"        // 错误消息
	MessageTypeHistory      MessageType = "history"      // 历史消息请求/响应
	MessageTypeAgentResult  MessageType = "agent_result" // Agent 执行结果（包含工具调用轨迹）
)

// Message WebSocket 消息
type Message struct {
	ID         string      `json:"id"`          // 消息唯一标识
	Type       MessageType `json:"type"`        // 消息类型
	From       string      `json:"from"`        // 发送者ID
	To         string      `json:"to"`          // 接收者ID（可选）
	SessionID  string      `json:"session_id"`  // 会话ID
	Timestamp  int64       `json:"timestamp"`   // 时间戳
	Content    interface{} `json:"content"`     // 消息内容
	ClientType int         `json:"client_type,omitempty"` // 客户端类型(1:App,2:Agent)
}

// NewMessage 创建新消息
func NewMessage(msgType MessageType, from, sessionID string, content interface{}) *Message {
	return &Message{
		ID:        generateMessageID(),
		Type:      msgType,
		From:      from,
		SessionID: sessionID,
		Timestamp: time.Now().Unix(),
		Content:   content,
	}
}

// ToJSON 序列化为JSON
func (m *Message) ToJSON() ([]byte, error) {
	return json.Marshal(m)
}

// FromJSON 从JSON反序列化
func (m *Message) FromJSON(data []byte) error {
	// 使用临时结构体来解析，避免直接解析 Content
	type tempMessage struct {
		ID        string          `json:"id"`
		Type      MessageType     `json:"type"`
		From      string          `json:"from"`
		To        string          `json:"to"`
		SessionID string          `json:"session_id"`
		Timestamp int64           `json:"timestamp"`
		Content   json.RawMessage `json:"content"`
	}

	var temp tempMessage
	if err := json.Unmarshal(data, &temp); err != nil {
		return err
	}

	m.ID = temp.ID
	m.Type = temp.Type
	m.From = temp.From
	m.To = temp.To
	m.SessionID = temp.SessionID
	m.Timestamp = temp.Timestamp

	// 根据消息类型解析 Content
	m.Content = parseContent(temp.Type, temp.Content)

	return nil
}

// parseContent 根据消息类型解析内容
func parseContent(msgType MessageType, rawContent json.RawMessage) interface{} {
	if len(rawContent) == 0 {
		return nil
	}

	switch msgType {
	case MessageTypeText:
		var content TextContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	case MessageTypeFile:
		var content FileContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	case MessageTypeToolCall:
		var content ToolCallContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	case MessageTypeToolResult:
		var content ToolResultContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	case MessageTypeStatus:
		var content StatusContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	case MessageTypeHeartbeat:
		var content HeartbeatContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	case MessageTypeHistory:
		var content HistoryContent
		if err := json.Unmarshal(rawContent, &content); err == nil {
			return &content
		}
	}

	// 如果解析失败，返回原始 map
	var generic map[string]interface{}
	if err := json.Unmarshal(rawContent, &generic); err == nil {
		return generic
	}

	return nil
}

// GetTextContent 安全获取文本内容
func (m *Message) GetTextContent() *TextContent {
	switch c := m.Content.(type) {
	case *TextContent:
		return c
	case map[string]interface{}:
		if text, ok := c["text"].(string); ok {
			return &TextContent{Text: text}
		}
	}
	return nil
}

// GetStatusContent 安全获取状态内容
func (m *Message) GetStatusContent() *StatusContent {
	switch c := m.Content.(type) {
	case *StatusContent:
		return c
	case map[string]interface{}:
		status, _ := c["status"].(string)
		message, _ := c["message"].(string)
		return &StatusContent{Status: status, Message: message}
	}
	return nil
}

// TextContent 文本消息内容
type TextContent struct {
	Text string `json:"text"` // 文本内容
}

// FileContent 文件消息内容
type FileContent struct {
	Name     string `json:"name"`     // 文件名
	Type     string `json:"type"`     // 文件类型
	Size     int64  `json:"size"`     // 文件大小
	Content  string `json:"content"`  // 文件内容（Base64）
}

// ToolCallContent 工具调用内容
type ToolCallContent struct {
	ToolName   string                 `json:"tool_name"`   // 工具名称
	Parameters map[string]interface{} `json:"parameters"`  // 工具参数
}

// ToolResultContent 工具结果内容
type ToolResultContent struct {
	ToolName string      `json:"tool_name"` // 工具名称
	Result   interface{} `json:"result"`    // 工具结果
	Error    string      `json:"error"`     // 错误信息
}

// StatusContent 状态消息内容
type StatusContent struct {
	Status  string `json:"status"`  // 状态
	Message string `json:"message"` // 状态描述
}

// HeartbeatContent 心跳消息内容
type HeartbeatContent struct {
	Timestamp int64 `json:"timestamp"` // 时间戳
}

// HistoryContent 历史消息内容
type HistoryContent struct {
	SessionID string                   `json:"session_id"` // 会话ID
	Messages  []map[string]interface{} `json:"messages"`   // 消息列表
	Total     int                      `json:"total"`      // 消息总数
}

// generateMessageID 生成消息ID
func generateMessageID() string {
	return time.Now().Format("20060102150405") + randomString(8)
}

// randomString 生成随机字符串
func randomString(n int) string {
	const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[i%len(letters)]
	}
	return string(b)
}

File: android-app/app/src/main/java/com/example/chatai/data/model/Message.kt
package com.example.chatai.data.model

import com.google.gson.annotations.SerializedName

/**
 * WebSocket消息模型
 */
data class Message(
    @SerializedName("id")
    val id: String,

    @SerializedName("type")
    val type: String,

    @SerializedName("from")
    val from: String,

    @SerializedName("to")
    val to: String? = null,

    @SerializedName("session_id")
    val sessionId: String,

    @SerializedName("timestamp")
    val timestamp: Long,

    @SerializedName("content")
    val content: MessageContent,

    /**
     * 客户端类型：1=用户发送的消息（右侧），2=Agent回复（左侧）
     * 用于判断消息展示方向
     */
    @SerializedName("client_type")
    val clientType: Int? = null
) {
    /**
     * 判断是否是自己发送的消息
     * 优先使用 client_type 判断：1=用户发送，2=Agent回复
     * 兜底使用 from 字段判断
     */
    fun isSentByMe(currentClientId: String): Boolean {
        // 优先根据 client_type 判断方向
        clientType?.let { return it == 1 }
        // 兜底：根据 from 判断
        return from == currentClientId
    }

    /**
     * 获取消息文本内容
     */
    fun getTextContent(): String? {
        return content.text
    }

    /**
     * 获取状态内容
     */
    fun getStatusContent(): String? {
        return content.message ?: content.status
    }

    /**
     * 获取显示内容 - 根据消息类型返回可读文本
     */
    fun getDisplayContent(): String {
        return when (type) {
            MessageType.TEXT -> content.text ?: ""
            MessageType.STATUS -> {
                val status = content.status ?: ""
                val msg = content.message ?: ""
                if (msg.isNotEmpty()) "$status: $msg" else status
            }
            MessageType.TOOL_CALL -> {
                val toolName = content.toolName ?: ""
                val params = content.parameters ?: ""
                "🔧 $toolName\n$params"
            }
            MessageType.TOOL_RESULT -> {
                val toolName = content.toolName ?: ""
                val result = content.result ?: ""
                "📋 $toolName\n$result"
            }
            MessageType.ERROR -> "❌ ${content.message ?: content.text ?: "未知错误"}"
            MessageType.HEARTBEAT -> "💓 心跳"
            MessageType.FILE -> "📁 ${content.name ?: "文件"}"
            else -> content.text ?: ""
        }
    }
}

/**
 * 消息内容模型
 */
data class MessageContent(
    @SerializedName("text")
    val text: String? = null,

    @SerializedName("status")
    val status: String? = null,

    @SerializedName("message")
    val message: String? = null,

    @SerializedName("timestamp")
    val timestamp: Long? = null,

    // 工具调用字段
    @SerializedName("tool_name")
    val toolName: String? = null,

    @SerializedName("parameters")
    val parameters: String? = null,

    // 工具结果字段
    @SerializedName("result")
    val result: String? = null,

    // 文件字段
    @SerializedName("name")
    val name: String? = null,

    @SerializedName("size")
    val size: Long? = null,

    @SerializedName("content")
    val fileContent: String? = null
)

/**
 * 消息类型常量
 */
object MessageType {
    const val TEXT = "text"
    const val FILE = "file"
    const val TOOL_CALL = "tool_call"
    const val TOOL_RESULT = "tool_result"
    const val HEARTBEAT = "heartbeat"
    const val STATUS = "status"
    const val ERROR = "error"
}
